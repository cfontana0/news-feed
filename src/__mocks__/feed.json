{
  "type": "feed-item",
  "relationships": {
    "sender": {
      "data": {
        "type": "user",
        "id": "124"
      }
    },
    "feed-status": {
      "data": null
    },
    "feed-request": {
      "data": null
    },
    "feed-message": {
      "data": {
        "type": "feed-message",
        "id": "1262"
      }
    }
  },
  "id": "1262",
  "attributes": {
    "updated-at": "2018-06-13T15:27:47Z",
    "sent-as": 124,
    "inserted-by": 124,
    "inserted-at": "2018-06-13T15:27:47Z",
    "id": 1262,
    "feed-status-id": null,
    "feed-request-id": null,
    "feed-message-id": 1262,
    "case-id": 162
  },
  "sender": {
    "type": "user",
    "id": "124",
    "attributes": {
      "updated-at": "2018-06-01T21:26:40Z",
      "name": "Austin Jackson",
      "inserted-at": "2018-06-01T21:26:40Z",
      "id": 124,
      "firm-user-id": 13,
      "email": "austin@example.com",
      "confirmed-at": "2018-06-01T21:27:08.029536Z",
      "client-user-id": null
    }
  },
  "message": {
    "type": "feed-message",
    "relationships": {
      "feed-item": {}
    },
    "id": "1262",
    "attributes": {
      "updated-at": "2018-06-13T15:27:47Z",
      "text": "We have received the following document: Lab Work from Crescent Labs, LLC.\n",
      "inserted-at": "2018-06-13T15:27:47Z",
      "id": 1262,
      "feed-message-type": null
    }
  }
}