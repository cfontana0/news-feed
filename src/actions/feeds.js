import { getFeeds } from '../utils/api'
import * as actionTypes from '../constants/actionTypes'

const onFeedsActionSuccess = (feeds) => {
  return {
    type: actionTypes.GET_FEEDS_SUCCESS,
    payload: {feeds}
  }
}

const onFeedsActionError = () => {
  return {
    type: actionTypes.GET_FEEDS_ERROR,
    payload: {}
  }
}

export const feedsAction = () => {
  return async (dispatch, getState) => {
    try {
      const req = await getFeeds()

      if (req.status === 200) {
        const feeds = req.data.data.map(feed => {
          const sender = req.data.included.filter(data => { return (data.type === 'user' && data.id === feed.relationships.sender.data.id) })[0]
          const message = req.data.included.filter(data => { return (data.type === 'feed-message' && data.id === feed.relationships['feed-message'].data.id) })[0]
          return {...feed, sender, message}
        })

        return dispatch(onFeedsActionSuccess(feeds))
      } else {
        return dispatch(onFeedsActionError())
      }
    } catch (e) {
      return dispatch(onFeedsActionError())
    }
  }
}
