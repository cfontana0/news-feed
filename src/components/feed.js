import React, { Component } from 'react'
import { Feed } from 'semantic-ui-react'
import moment from 'moment'

class FeedItem extends Component {
  render () {
    const { feed } = this.props

    return (
      <Feed size='large'>
        <Feed.Event>
          <Feed.Label>{ feed.sender.attributes.name.charAt(0) }</Feed.Label>
          <Feed.Content>
            <Feed.Summary>
              <Feed.User>{ feed.sender.attributes.name }</Feed.User> added a new feed
            </Feed.Summary>
            <Feed.Meta>
              <Feed.Date>{ moment(feed.attributes['inserted-at']).fromNow() }</Feed.Date>
            </Feed.Meta>
            <Feed.Extra text>
              { feed.message.attributes.text }
            </Feed.Extra>
          </Feed.Content>
        </Feed.Event>
      </Feed>
    )
  }
}

export default FeedItem
