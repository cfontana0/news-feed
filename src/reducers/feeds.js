import * as actionTypes from '../constants/actionTypes'

const initialState = {
  form: {
    list: []
  }
}

export default function (state = initialState, action) {
  switch (action.type) {
    case actionTypes.GET_FEEDS_SUCCESS:
      return {...state, list: action.payload.feeds}
    default:
      return state
  }
}
