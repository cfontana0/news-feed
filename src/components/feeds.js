import React, { Component } from 'react'
import { feedsAction } from '../actions/feeds'
import { connect } from 'react-redux'
import { bindActionCreators } from 'redux'
import Feed from './feed'
import { Container, Header } from 'semantic-ui-react'

export class FeedsComponent extends Component {
  componentWillMount () {
    this.props.actions.feedsAction()
  }

  render () {
    const feeds = (this.props.feeds || []).map(feed => {
      return (<Feed key={feed.id} feed={feed}/>)
    })

    return (
      <div className='main'>
        <Container fluid>
          <Header as='h2'>Feeds & News</Header>
          { (!this.props.feeds) && <span>Loading..</span> }
          { feeds }
        </Container>
      </div>
    )
  }
}

const mapStateToProps = ({ feeds }) => ({
  feeds: feeds.list
})

const mapDispatchToProps = dispatch => ({
  actions: bindActionCreators({ feedsAction }, dispatch)
})

export default connect(mapStateToProps, mapDispatchToProps)(FeedsComponent)
