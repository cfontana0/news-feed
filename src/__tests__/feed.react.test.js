import Feed from '../components/feed';
import { FeedsComponent } from '../components/feeds';
import React from 'react';
import renderer from 'react-test-renderer';
import feed from '../__mocks__/feed'

describe('Feed component renders the feed correctly', () => {
  it('renders correctly', () => {

    const rendered = renderer.create(
      <Feed feed={feed} />
    );

    expect(rendered.toJSON()).toMatchSnapshot();
  });
});

describe('Feeds list component renders the feeds correctly', () => {
  it('renders correctly', () => {

    const actions = {
      feedsAction: () => {}
    }

    const rendered = renderer.create(
      <FeedsComponent actions={actions} feeds={[feed]}/>
    );

    expect(rendered.toJSON()).toMatchSnapshot();
  });
});