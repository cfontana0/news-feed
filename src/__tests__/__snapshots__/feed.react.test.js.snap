// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Feed component renders the feed correctly renders correctly 1`] = `
<div
  className="ui large feed"
>
  <div
    className="event"
  >
    <div
      className="label"
    >
      A
    </div>
    <div
      className="content"
    >
      <div
        className="summary"
      >
        <a
          className="user"
        >
          Austin Jackson
        </a>
         added a new feed
      </div>
      <div
        className="meta"
      >
        <div
          className="date"
        >
          a month ago
        </div>
      </div>
      <div
        className="text extra"
      >
        We have received the following document: Lab Work from Crescent Labs, LLC.
        
      </div>
    </div>
  </div>
</div>
`;

exports[`Feeds list component renders the feeds correctly renders correctly 1`] = `
<div
  className="main"
>
  <div
    className="ui fluid container"
  >
    <h2
      className="ui header"
    >
      Feeds & News
    </h2>
    <div
      className="ui large feed"
    >
      <div
        className="event"
      >
        <div
          className="label"
        >
          A
        </div>
        <div
          className="content"
        >
          <div
            className="summary"
          >
            <a
              className="user"
            >
              Austin Jackson
            </a>
             added a new feed
          </div>
          <div
            className="meta"
          >
            <div
              className="date"
            >
              a month ago
            </div>
          </div>
          <div
            className="text extra"
          >
            We have received the following document: Lab Work from Crescent Labs, LLC.
            
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
